### PR TITLE
Use type switch to extract message

### PR DESCRIPTION
**Pull request by ethanfrey**
- [x] Add ExtractMsgFromSum to reflect the embedded message
- [x] Pass bnsd tests
- [x] Pull logic into top-level app dir
- [x] Use logic in other apps
- [x] Clean up errors

### CODE REVIEW

**Comment by ruseinov** [comment]
This PR looks good, going to leave a couple of nits for errors

File: tx.go [hunk 1]
```diff
@@ -67,3 +74,32 @@ func GetPath(tx Tx) string {
...
+	if sum == nil {
+		return nil, errors.ErrInternal("sum is <nil>")
+	}
+	pval := reflect.ValueOf(sum)
+	if pval.Kind() != reflect.Ptr || pval.Elem().Kind() != reflect.Struct {
+		return nil, errors.ErrInternal(fmt.Sprintf("invalid value: %T", sum))
+	}
+	val := pval.Elem()
+	if val.NumField() != 1 {
+		return nil, errors.ErrInternal(fmt.Sprintf("Unexpected field count: %d", val.NumField()))
```
Comment (ruseinov) [inline tx.go]: Unexpected field count for transaction type?

Comment (ethanfrey) [reply tx.go]: Read godoc for the function.
It expects to have a certain format, that all protobuf sum types are....

```
&Sum_type{
  tx_xyz: &MsgImplementation{....}
}
```

And we just ensure this is the case before blindly grabbing fields

File: tx.go [hunk 2]
```diff
@@ -67,3 +74,32 @@ func GetPath(tx Tx) string {
...
+//
+//   ExtractMsgFromSum(tx.GetSum())
+//
+// To work, this requires sum to be a pointer to a struct with one field,
+// and that field can be cast to a Msg.
+// Returns an error if it cannot succeed.
+func ExtractMsgFromSum(sum interface{}) (Msg, error) {
+	// TODO: add better error messages here with new refactor
+	if sum == nil {
+		return nil, errors.ErrInternal("sum is <nil>")
```
Comment (ruseinov) [inline tx.go]: Sum bothers me in general, it does not self document. As this is basically transaction value aka it's concrete type. Might not be a concern for this PR, just something that I noticed. Seeing "sum is <nil>" in the log is not very descriptive, but at the same time this error should never happen. So let's leave it at that, but I'd think about changing this naming to something more self-documenting.

File: tx.go (hunk 1 above)
Comment (ruseinov) [reply tx.go]: @ethanfrey this was an error message suggestion, I didn't express it well enough.

**Pull request merged**